File: tempestweb/native/quota.py
"""Native storage-quota capability over the StorageManager Web API.

:func:`estimate` sends a ``quota.estimate`` ``native_call`` and :func:`persist` /
:func:`persisted` manage durable storage; ``client/native/quota.js`` calls
``navigator.storage.estimate`` / ``navigator.storage.persist`` /
``navigator.storage.persisted``. Persistent storage exempts the origin's data from
eviction under storage pressure.
"""

from __future__ import annotations

from dataclasses import dataclass

from tempestweb.native.dispatch import send_native_call

__all__ = ["StorageEstimate", "estimate", "persist", "persisted"]


@dataclass(frozen=True)
class StorageEstimate:
    """An estimate of the origin's storage usage and quota.

    Attributes:
        usage: Bytes currently used by the origin.
        quota: Total bytes available to the origin.
    """

    usage: int
    quota: int


async def estimate() -> StorageEstimate:
    """Estimate the origin's storage usage and quota.

    Returns:
        The current :class:`StorageEstimate`.

    Raises:
        NativeError: If the StorageManager API is unavailable (``unavailable``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("quota.estimate", {})
    return StorageEstimate(
        usage=int(value.get("usage", 0)),
        quota=int(value.get("quota", 0)),
    )


async def persist() -> bool:
    """Request that the origin's storage be made persistent.

    Returns:
        ``True`` if storage is persistent after the request.

    Raises:
        NativeError: If the API is unavailable (``unavailable``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("quota.persist", {})
    return bool(value.get("persisted", False))


async def persisted() -> bool:
    """Report whether the origin's storage is already persistent.

    Returns:
        ``True`` if storage is persistent.

    Raises:
        NativeError: If the API is unavailable (``unavailable``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("quota.persisted", {})
    return bool(value.get("persisted", False))

File: client/native/quota.js
// native/quota.js — StorageManager (quota + persistence) glue.
//
// `navigator.storage.estimate()` reports approximate usage/quota; `persist()`
// asks the browser to make origin storage durable (survives eviction pressure);
// `persisted()` reports whether it already is — without prompting.

import { CapabilityError } from "./index.js";

/**
 * Estimate current storage usage and quota (bytes).
 * @param {Object} _args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{usage:number, quota:number}>}
 * @throws {CapabilityError} unavailable when the StorageManager API is missing.
 */
export async function quotaEstimate(_args, deps) {
  const nav = deps.navigator || /** @type {any} */ (globalThis).navigator;
  if (!nav || !nav.storage || typeof nav.storage.estimate !== "function") {
    throw new CapabilityError("unavailable", "the StorageManager API is not available");
  }
  const est = await nav.storage.estimate();
  return { usage: est.usage || 0, quota: est.quota || 0 };
}

/**
 * Request durable (persistent) storage for this origin.
 * @param {Object} _args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{persisted:boolean}>}
 * @throws {CapabilityError} unavailable when the StorageManager API is missing.
 */
export async function quotaPersist(_args, deps) {
  const nav = deps.navigator || /** @type {any} */ (globalThis).navigator;
  if (!nav || !nav.storage || typeof nav.storage.persist !== "function") {
    throw new CapabilityError("unavailable", "the StorageManager API is not available");
  }
  return { persisted: await nav.storage.persist() };
}

/**
 * Report whether origin storage is already persistent.
 * @param {Object} _args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{persisted:boolean}>}
 * @throws {CapabilityError} unavailable when the StorageManager API is missing.
 */
export async function quotaPersisted(_args, deps) {
  const nav = deps.navigator || /** @type {any} */ (globalThis).navigator;
  if (!nav || !nav.storage || typeof nav.storage.persisted !== "function") {
    throw new CapabilityError("unavailable", "the StorageManager API is not available");
  }
  return { persisted: await nav.storage.persisted() };
}
